### ISSUE TITLE
async wait/sleep implementation

### ISSUE THREAD

**Issue by russkel** [ros2/rclpy#1234]
#### Feature description
I have been writing some hardware "test" scripts used for checking the integration of hardware in real time (not using any test framework or assertions but simply stimulating a system with inputs and manually confirming the hardware is performing as expected). I have found writing this using async has made things much simpler and nicer:

e.g.

```python
    async def test_1(self):
        self.get_logger().info("Setting Rudder to Center")
        await self.set_and_wait_until_rudder_angle(0.0, 5.0)

        self.get_logger().info("Setting Rudder to Full Port")
        await self.set_and_wait_until_rudder_angle(-1.0, 5.0)
```

To which I kick off like so:

```python
def main(args=None):
    rclpy.init(args=args)
    executor = SingleThreadedExecutor()
    node = TestNode()
    executor.add_node(node)

    try:
        node.get_logger().info("Starting Test 1")
        task = executor.create_task(node.test_1())
        executor.spin_until_future_complete(task)

        node.get_logger().info("Testing complete!")
    except KeyboardInterrupt:
        pass

    node.destroy_node()
    executor.shutdown()
```

Now there is pretty much no examples of this `create_task` being used like this so I have been experimenting and reading a bit of the rclpy code. I wasn't sure how to use the executor to do a `await asyncio.sleep` after looking at the code in depth. I had a quick look at how `asyncio` did it and I came up with this hack using a `Timer`:

```python
    async def async_wait_for(self, rel_time: float):
        fut = Future()
        timer = None

        def done_waiting():
            fut.set_result(None)
            timer.cancel()

        timer = self.create_timer(rel_time, done_waiting)
        await fut
```

This was the only missing piece and I was able to asynchronously "wait" and write the co-routines as expected.

My request is something like this is added to the `Executor` class, perhaps using a more elegant interface than abusing a Timer like this.

#279 might be related.

**Comment by arjo129** [comment]
I like this idea. I recently ran into a similar need. Perhaps `spin_until_future_complete` has a timeout option which I abused with a default empty future that never gets set.

**Comment by russkel** [comment]
> I like this idea. I recently ran into a similar need. My solution was `spin_until_future_complete` has a timeout option which I abused with a default empty future that never gets set. It probably is not great but 🤷.

If I am not mistaken that will block the thread you're in? How do you use that to await in a co-routine?

**Comment by arjo129** [comment]
Yeah it does. I wasnt using async await so... If an async/await option were available I'd probably have prefered that.

**Comment by russkel** [comment]
> Yeah it does. I wasnt using async await so... If an async/await option were available I'd probably have prefered that. Just echoing the sentiment that an API like this would be useful for people.

Yep I understood what you meant, I was just wondering if I had missed something.

**Comment by jmblixt3** [comment]
> > I like this idea. I recently ran into a similar need. My solution was `spin_until_future_complete` has a timeout option which I abused with a default empty future that never gets set. It probably is not great but 🤷.
> 
> If I am not mistaken that will block the thread you're in? How do you use that to await in a co-routine?

I was also looking into this, and looks like since #1316  `spin_until_future_complete` now works in callbacks. I'm still not a fan of relying on an empty future never finishing though

**Comment by Flova** [comment]
> Now there is pretty much no examples of `create_task` being used like this so I have been experimenting and reading a bit of the rclpy code. After looking at the executor code in depth I couldn't find out how to use the executor to do an equivalent to `await asyncio.sleep`. I had a quick look at how `asyncio` implemented theirs and I came up with this hack using a `Timer`:
> 
>     async def async_wait_for(self, rel_time: float):
>         fut = Future()
>         timer = None
> 
>         def done_waiting():
>             fut.set_result(None)
>             timer.cancel()
> 
>         timer = self.create_timer(rel_time, done_waiting)
>         await fut
> 
> This was the only missing piece and I was able to asynchronously "wait" and write the co-routines as expected.


Thanks for the provided snipped @russkel . Without async/await rclpy is not really usable (without hacks) due to deadlocks imo.. Together with the new EventsExecutor they also improve the performance a lot. Many of our nodes are down from 80% using the MultiThreadedExecutor to ~3% using the EventsExecutor + async.  

I noticed a few flaws in the provided snipped, that resulted in quite significant performance issues / resource leaks. This is my modified version that does not leak any resources as far as my testing goes:

```py
async def async_wait_for(node: Node, rel_time: float):
    future = Future()
    rel_time = max(rel_time, 0.0)

    def done_waiting():
        future.set_result(None)

    timer = node.create_timer(rel_time, done_waiting, clock=node.get_clock())
    await future
    timer.cancel()
    node.destroy_timer(timer)
```

**Comment by russkel** [comment]
This is a good update, thanks for that. I will update my comment pointing to yours. I've only used it in very short lived test scripts (and hastily put together) so I wasn't thinking about this, but yes it would create a lot of timers..

**Comment by russkel** [comment]
I still would like this to be formalised. @clalancette does this make sense to be added?

**Comment by Yadunund** [comment]
@sloretz please take a look when you get a chance 🧇

**Comment by russkel** [comment]
Presumably with the async executor this will no longer be necessary?
